Index: lib/System/Win32/DynamicLibrary.inc
===================================================================
--- lib/System/Win32/DynamicLibrary.inc	(revision 97727)
+++ lib/System/Win32/DynamicLibrary.inc	(working copy)
@@ -118,7 +118,7 @@
   #define EXPLICIT_SYMBOL_DEF(SYM)                \
     extern "C" { extern void *SYM; }
 
-  #if defined(__MINGW32__)
+  #if defined(__MINGW32__) && !defined(__MINGW64__)
     EXPLICIT_SYMBOL_DEF(_alloca);
     EXPLICIT_SYMBOL_DEF(__main);
     EXPLICIT_SYMBOL_DEF(__ashldi3);
@@ -159,7 +159,7 @@
     }
   }
 
-#if defined(__MINGW32__)
+#if defined(__MINGW32__) && !defined(__MINGW64__)
   {
     EXPLICIT_SYMBOL(_alloca);
     EXPLICIT_SYMBOL(__main);
